### PR TITLE
rv-virt:knsh64

### PR DESCRIPTION
**Pull request by masayuki2009**
## Summary

- This PR consists of the following 2 commits
- commit1: arch: risc-v: Add MMU support for qemu-rv
- commit2: boards: rv-virt: Add knsh64 related files

## Impact

- qemu-rv only

## Testing

- Tested with qemu-6.2
- Please note that mtimer is disabled for S-mode because
  the mtimer needs to be accessed in M-mode

### CODE REVIEW

**Comment by pussuw** [comment]
Where are the M-mode traps directed to ?

File: arch/risc-v/src/qemu-rv/qemu_rv_timerisr.c [hunk 1]
```diff
@@ -61,11 +61,13 @@
 
 void up_timer_initialize(void)
 {
+#ifndef CONFIG_BUILD_KERNEL
```
Comment (pussuw) [inline arch/risc-v/src/qemu-rv/qemu_rv_timerisr.c:64]: I think this is unnecessary. The mtimer driver forwards access to SBI, if CONFIG_ARCH_USE_S_MODE is defined.

But this is fine also if you don't need the ticker.

**Comment by masayuki2009** [comment]
>Where are the M-mode traps directed to ?

@pussuw 
The M-mode traps are still routed to __trap_vec in qemu_rv_head.S
However, I think this should be fixed later if S-mode is enabled.

File: arch/risc-v/src/qemu-rv/qemu_rv_timerisr.c (hunk 1 above)
Comment (masayuki2009) [reply arch/risc-v/src/qemu-rv/qemu_rv_timerisr.c:64]: The SBI is not integrated yet, so the mtimer is disabled.

**Comment by pussuw** [comment]
> > Where are the M-mode traps directed to ?
> 
> @pussuw The M-mode traps are still routed to __trap_vec in qemu_rv_head.S However, I think this should be fixed later if S-mode is enabled.

Ok, the SBI part of BUILD_KERNEL is still work in progress, when it is done I can fixup this too.

**Review by pussuw — APPROVED** [review]
LGTM

**Comment by masayuki2009** [comment]
@pussuw 
Thanks for reviewing this PR.

**Comment by masayuki2009** [comment]
```
curl: (28) Failed to connect to www.phyplusinc.com port 80: Connection timed out
cp: cannot stat 'libphy62xxble.a': No such file or directory
make[1]: *** [chip/Make.defs:82: .buildlib] Error 1
make[1]: Target 'context' not remade because of errors.
make: *** [tools/Unix.mk:425: arch/arm/src/.context] Error 2
make: Target 'all' not remade because of errors.
```

**Comment by GooTal** [comment]
Hello, i tried to run rv-virt:knsh64. But it didn`t work well. Here are the commands:
```
./tools/configure.sh -l rv-virt:knsh64   # step1
make -j8   # step2
qemu-system-riscv64 -semihosting -M virt -cpu rv64 -smp 8 -bios none -kernel nuttx -nographic  #step3
```
 I didn`t got any error in step1 and step2. But the error occurred in step3 and it`s stuck here:
 
图片

Did i use the right command for qemu-system-riscv64 to run nuttx? Or did i miss something else?

Thanks a lot!
@xiaoxiang781216 @masayuki2009

**Comment by xiaoxiang781216** [comment]
@masayuki2009 could you help?

**Comment by masayuki2009** [comment]
@GooTal 
Please see boards/risc-v/qemu-rv/rv-virt/README.txt

```
3.2 Configure and build NuttX for BUILD_KERNEL

  $ mkdir ./nuttx; cd ./nuttx
  $ git clone https://github.com/apache/nuttx.git nuttx
  $ git clone https://github.com/apache/nuttx-apps.git apps
  $ cd nuttx
  $ make distclean
  $ ./tools/configure.sh rv-virt:knsh64
  $ make V=1 -j7
  $ make export V=1
  $ cd ../apps
  $ ./tools/mkimport.sh -z -x ../nuttx/nuttx-export-*.tar.gz
  $ make import V=1
  $ cd ../nuttx
```

**Comment by davids5** [comment]
By "added" did you run `make menuconfig` or add them by hand?  `make menuconfig` is needed to insure the dependency.

**Comment by masayuki2009** [comment]
@GooTal 

>BTW, can knsh64 run ostest? I notice that rv-virt:nsh64 got some built-in apps:

Currently, no. 
Because the ostest uses some non-POISX API (i.e. task_create) which can not be used for BUILD_KERNEL.

**Comment by GooTal** [comment]
> By "added" did you run `make menuconfig` or add them by hand? `make menuconfig` is needed to insure the dependency.

Oh sorry. I added them by hand. Thanks @davids5. It\`s really helpful that using menuconfig to ensure the dependency.

And thanks a lot @masayuki2009 . I just noticed this in `nuttx/libs/libc/builtin/Kconfig` so that the menuconfig do not show this option.
```
config BUILTIN
	bool "Support Builtin Applications"
	default n
	depends on !BUILD_KERNEL
```

**Comment by xiaoxiang781216** [comment]
There is a draft PR to enable ostest for kernel mode: https://github.com/apache/nuttx-apps/pull/1500. @GooTal you can try and refine it.

**Comment by GooTal** [comment]
> There is a draft PR to enable ostest for kernel mode: [apache/nuttx-apps#1500](https://github.com/apache/nuttx-apps/pull/1500) from @pussuw . @GooTal you can try and refine it.

Thanks for your invitation. I\`ll follow the updates during porting NuttX to LoongArch. Currently I\`m porting nsh64 and knsh64 to qemu-system-loongarch64.

**Pull request merged**